op-node/p2p/pings.go

```diff
@@ -0,0 +1,134 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	"golang.org/x/time/rate"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-service/clock"
+)
+
+const (
+	pingRound                 = 10 * time.Minute
+	pingsPerSecond rate.Limit = 1
+	pingsBurst                = 10
+)
+
+type PingFn func(ctx context.Context, peerID peer.ID) <-chan ping.Result
+
+type PeersFn func() []peer.ID
+
+type PingService struct {
+	ping  PingFn
+	peers PeersFn
+
+	clock clock.Clock
+
+	log log.Logger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	trace func(work string)
+
+	// to signal service completion
+	wg sync.WaitGroup
+}
+
+func NewPingService(log log.Logger, ping PingFn, peers PeersFn, clock clock.Clock) *PingService {
+	ctx, cancel := context.WithCancel(context.Background())
+	srv := &PingService{
+		ping:   ping,
+		peers:  peers,
+		log:    log,
+		clock:  clock,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	srv.wg.Add(1)
+	go srv.pingPeersBackground()
+	return srv
+}
+
+func (p *PingService) Close() {
+	p.cancel()
+	p.wg.Wait()
+}
+
+func (e *PingService) pingPeersBackground() {
+	defer e.wg.Done()
+
+	tick := e.clock.NewTicker(pingRound)
+	defer tick.Stop()
+
+	if e.trace != nil {
+		e.trace("started")
+	}
+
+	for {
+		select {
+		case <-tick.Ch():
+			e.pingPeers()
+		case <-e.ctx.Done():
+			return
+		}
+	}
+}
+
+func (e *PingService) pingPeers() {
+	if e.trace != nil {
+		e.trace("pingPeers start")
+	}
+	ctx, cancel := context.WithTimeout(e.ctx, pingRound)
+	defer cancel()
+
+	// Wait group to wait for all pings to complete
+	var wg sync.WaitGroup
+	// Rate-limiter to help schedule the ping
+	// work without overwhelming ourselves.
+	rl := rate.NewLimiter(pingsPerSecond, pingsBurst)
+
+	// iterate through the connected peers
+	for i, peerID := range e.peers() {
+		if e.ctx.Err() != nil { // stop if the service is closing or timing out
+			return
+		}
+		if ctx.Err() != nil {
+			e.log.Warn("failed to ping all peers", "pinged", i, "err", ctx.Err())
+			return
+		}
+		if err := rl.Wait(ctx); err != nil {
+			// host may be getting closed, causing a parent ctx to close.
+			return
+		}
+		wg.Add(1)
+		go func(peerID peer.ID) {
+			e.pingPeer(ctx, peerID)
+			wg.Done()
+		}(peerID)
+	}
```

Comment:
Because all peers are pinged at the same time, all the results will be affected by the local performance at the time the ping is taken. That is probably okay, but if we spread out this loop so that different peers are collected at different times, it'd reduce that.

Reply:
The `rl.Wait(ctx)` call already performs a rate-limit, so we're not pinging too many peers at once. Using a rate-limit this way is easier than trying to organize the peers in parallel sub-sets.